### ISSUE TITLE
gp(solvers): implement dense solver surface and gaussx-backed linear algebra bridge

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#21]
Source draft ID: `PYX-16`

The wave-level design context is copied below so this issue stays self-contained for another implementing agent.

## Shared Context
This grouped draft is intentionally carrying the GP design context inline so the Wave 2 issue set stands on its own. The relevant source docs were `vision.md`, `architecture.md`, `api/gp/primitives.md`, `api/gp/components.md`, `api/gp/models.md`, `examples/gp/primitives.md`, and `features/gp/models.md`.

The GP stack is organized as layers:

- Layer 0: pure primitives such as kernels, covariance algebra, whitening, Kalman/quadrature helpers.
- Layer 1: protocols and components such as `Kernel`, `Solver`, `Guide`, `Integrator`, and `Likelihood`.
- Layer 2: model-facing entry points such as `GPPrior`, `ConditionedGP`, `gp_sample`, and `gp_factor`.
- Later layers: sparse guides, inference strategies, state-space models, pathwise samplers, and multi-output structure.

Wave 2 only ships the dense foundation. It should not silently drift into sparse, temporal, or multi-output work.

## Core Modeling Principle
The design docs make one principle explicit: the user owns the NumPyro model, while `pyrox.gp` owns the GP math. That means Wave 2 should already support both of the canonical GP entry points:

- `gp_factor("gp", prior, y, noise_var)` for collapsed Gaussian-likelihood workflows.
- `gp_sample("f", prior, guide=...)` for latent-function workflows under non-conjugate likelihoods.

Those two entry points are the backbone for later waves, so Wave 2 docs and tests should teach them clearly.

## Mathematical Baseline
The foundational formulas that later waves build on are:

Collapsed marginal likelihood:

$$\log p(y \mid X, \theta) = -\tfrac{1}{2} y^\top (K + \sigma^2 I)^{-1} y - \tfrac{1}{2}\log |K + \sigma^2 I| - \tfrac{N}{2}\log(2\pi)$$

Posterior predictive moments:

$$\mu_* = K_{*f}(K_{ff} + \sigma^2 I)^{-1} y$$

$$\Sigma_* = K_{**} - K_{*f}(K_{ff} + \sigma^2 I)^{-1} K_{f*}$$

Whitening/projection identity reused later by sparse methods:

$$A_X = K_{XZ} K_{ZZ}^{-1} = L_Z^{-\top}(L_Z^{-1} K_{XZ}^\top)^\top$$

## Wave 2 Public Surface
If this wave lands correctly, another agent should be able to assume the following names are stable enough to build on:

- `RBF`, `Matern`, `Periodic`, `Linear`
- `Kernel`, `Solver`, `Guide`, `Integrator`, `Likelihood`
- `CholeskySolver`
- `GPPrior`, `ConditionedGP`
- `gp_factor`, `gp_sample`

## Explicit Non-Goals
Wave 2 should not quietly absorb:

- sparse guides
- natural-gradient or site-based inference strategies
- temporal/state-space GP workflows
- pathwise posterior samplers
- multi-output kernels

Those belong to later grouped drafts below.

---

## Problem / Request
Implement the Wave 2 dense solver surface and make it compatible with GaussX-backed solve/logdet workflows.

## User Story
As a GP model implementer, I want a solver object that exposes factorization, solve, and log-determinant behavior cleanly so `GPPrior` and posterior conditioning do not hard-code dense linear algebra inline.

## Motivation
Dense solve/logdet behavior is the first reusable systems boundary for `pyrox.gp`. Posterior conditioning, marginal likelihoods, and later sparse or structured solvers all depend on this line being clean.

## Proposed API
```python
class CholeskySolver(eqx.Module):
    jitter: float = 1e-6

    def factor(self, K): ...
    def solve(self, K, rhs): ...
    def logdet(self, K): ...
    def posterior_coefficients(self, K, y, noise_var): ...
```

## Mathematical Notes
Wave 2 only needs the dense Gaussian atoms that appear in every GP derivation:

$$\alpha = (K + \sigma^2 I)^{-1} y$$

$$\log |K + \sigma^2 I|$$

The issue should not broaden into iterative or structured solvers yet, but it should leave enough abstraction that later Kalman, Kronecker, or sparse solvers fit behind the same conceptual interface.

## References & Existing Code
- Design doc / spec: `design_docs/pyrox/api/gp/models.md`, `design_docs/pyrox/architecture.md`, `design_docs/pyrox/examples/integration.md`
- Existing code to compare against: the dense posterior formulas in `api/gp/models.md`
- Related prior art: Cholesky-based dense GP regression workflows

## Implementation Notes
- Implement a dense Cholesky-backed solver first.
- Use GaussX operators or conventions where helpful so the later structured-solver story stays aligned.
- Keep the public surface small and focused on factor, solve, logdet, and posterior coefficients.

## Implementation Steps
- [ ] Implement a dense Cholesky-backed solver first.
- [ ] Use GaussX operators or conventions where helpful so the later structured-solver story stays aligned.
- [ ] Keep the public surface small and focused on factor, solve, logdet, and posterior coefficients.

## Definition of Done
- Wave 2 can compute dense solves, log-determinants, and posterior coefficients through a solver object.
- The solver surface is stable enough for `GPPrior` and `ConditionedGP`.

## Testing
- Verify solves and logdets against direct `jax.numpy` linear algebra on small PSD systems.
- Verify posterior coefficient helpers against explicit solves.

## Documentation
- Document that Wave 2 solver support is intentionally dense-first; structured solvers arrive later.

## Relationships
- Parent epic: #17.
- Blocked by #20.
- Blocks #22.
